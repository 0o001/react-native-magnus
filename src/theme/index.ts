export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
