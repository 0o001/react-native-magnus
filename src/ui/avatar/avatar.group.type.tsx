import { ImageSourcePropType as RNImageSourcePropType } from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
  ShadowPropsType,
} from '../../theme';

export interface AvatarGroupProps
  extends BorderPropsType,
    SpacingPropsType,
    RoundedPropsType,
    ShadowPropsType {
  h?: number | string;
  w?: number | string;
  bg?: string;
  minH?: number;
  minW?: number;
  maxH?: number;
  maxW?: number;
  bgImg?: RNImageSourcePropType;
  bgMode?: 'contain' | 'cover' | 'stretch';
  flex?: number;
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  alignSelf?:
    | 'auto'
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'stretch'
    | 'baseline';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline';
  flexDir?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
  flexWrap?: 'wrap' | 'nowrap' | 'wrap-reverse';
  position?: 'absolute' | 'relative';
  overflow?: 'hidden' | 'visible' | 'scroll';
  borderStyle?: 'solid' | 'dotted' | 'dashed';
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
  opacity?: number;
  children: React.ReactElement[] | React.ReactElement;
  row?: boolean;
  zIndex?: number;
  offset?: string | number;
}
