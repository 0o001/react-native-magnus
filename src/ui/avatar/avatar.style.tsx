import { StyleSheet } from 'react-native';

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
  createBorderRadiusStyles,
  createPositionStyle,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.container = {
    backgroundColor: getThemeProperty(theme.colors, props.bg, 'transparent'),
    alignItems: 'center',
    justifyContent: 'center',
    ...createPositionStyle(props),
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  computedStyle.text = {
    color: getThemeProperty(theme.colors, props.color, '#000'),
    fontSize: getThemeProperty(theme.fontSize, props.fontSize, 16),
    textAlignVertical: 'center',
    textAlign: 'center',
  };

  computedStyle.image = {
    height: props.size,
    width: props.size,
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  if (props.flexDir) {
    computedStyle.container = {
      ...computedStyle.container,
      flexDirection: props.flexDir,
    };
  }

  if (props.size) {
    computedStyle.container = {
      ...computedStyle.container,
      width: props.size,
      height: props.size,
    };
  }

  if (props.shadow) {
    computedStyle.container = {
      ...computedStyle.container,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, 'white'),
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.container = {
      ...computedStyle.container,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
