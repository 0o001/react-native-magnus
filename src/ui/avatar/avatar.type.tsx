import {
  ViewProps as RNViewProps,
  ImageSourcePropType as RNImageSourcePropType,
} from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from '../../theme';

export interface AvatarProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType {
  size?: number;
  color?: string;
  bg?: string;
  source?: RNImageSourcePropType;
  fontSize?: string;
  shadow?: number;
  shadowColor?: string;
  position?: 'absolute' | 'relative';
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
  alignSelf?:
    | 'auto'
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'stretch'
    | 'baseline';
  zIndex?: number;
  opacity?: number;
}
