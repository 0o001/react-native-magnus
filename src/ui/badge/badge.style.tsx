import { StyleSheet } from 'react-native';

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
  createBorderRadiusStyles,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.container = {
    minHeight: typeof props.count === 'undefined' ? 10 : 35,
    minWidth: typeof props.count === 'undefined' ? 10 : 35,
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  computedStyle.div = {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    backgroundColor: getThemeProperty(theme.colors, props.bg, 'transparent'),
    ...createBorderRadiusStyles(props, theme.borderRadius),
    minHeight: typeof props.count === 'undefined' ? 10 : 35,
    minWidth: typeof props.count === 'undefined' ? 10 : 35,
    zIndex: 9999,
  };

  computedStyle.text = {
    color: props.color,
    fontSize: getThemeProperty(theme.fontSize, props.fontSize, 16),
    paddingHorizontal: 6,
  };

  if (props.shadow) {
    computedStyle.div = {
      ...computedStyle.div,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, 'white'),
    };
  }

  if (props.h) {
    computedStyle.div = {
      ...computedStyle.div,
      height: props.h,
    };
  }

  if (props.w) {
    computedStyle.div = {
      ...computedStyle.div,
      width: props.w,
    };
  }

  if (props.top) {
    computedStyle.div = {
      ...computedStyle.div,
      top: props.top,
    };
  }

  if (props.right) {
    computedStyle.div = {
      ...computedStyle.div,
      right: props.right,
    };
  }

  if (props.bottom) {
    computedStyle.div = {
      ...computedStyle.div,
      bottom: props.bottom,
    };
  }

  if (props.left) {
    computedStyle.div = {
      ...computedStyle.div,
      left: props.left,
    };
  }

  if (props.minH) {
    computedStyle.div = {
      ...computedStyle.div,
      minHeight: props.minH,
    };
  }

  if (props.minW) {
    computedStyle.div = {
      ...computedStyle.div,
      minWidth: props.minW,
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.div = {
      ...computedStyle.div,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
