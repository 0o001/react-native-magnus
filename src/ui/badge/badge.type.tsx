import {
  ViewProps as RNViewProps,
  ImageSourcePropType as RNImageSourcePropType,
} from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from '../../theme';

export interface BadgeProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType {
  h?: number | string;
  w?: number | string;
  bg?: string;
  minH?: number;
  minW?: number;
  shadow?: number;
  shadowColor?: string;
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline';
  position?: 'absolute' | 'relative';
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
  color?: string;
  fontSize?: string;
  count?: string | number;
  children?: React.ReactNode[] | React.ReactNode;
}
