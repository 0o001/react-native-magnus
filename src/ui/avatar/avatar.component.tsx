import * as React from 'react';
import { useContext } from 'react';
import { Text as RNText, View as RNView, Image as RNImage } from 'react-native';

import { getStyle } from './avatar.style';
import { ThemeContext } from '../../theme';
import { AvatarGroup } from './avatar.group.component';
import { AvatarProps, CompoundedAvatar } from './avatar.type';

const Avatar: CompoundedAvatar<AvatarProps> = (props) => {
  const {
    bg,
    m,
    mt,
    mr,
    mb,
    ml,
    ms,
    p,
    pr,
    pt,
    pb,
    pl,
    position,
    style,
    rounded,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    children,
    borderColor,
    color,
    fontSize,
    size,
    alignSelf,
    bottom,
    left,
    right,
    top,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderBottomWidth,
    borderTopWidth,
    borderEndWidth,
    shadow,
    shadowColor,
    opacity,
    zIndex,
    source,
    ...rest
  } = props;
  const { theme } = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  /**
   * render the content based on the props passed
   */
  const renderContent = () => {
    if (source) {
      return <RNImage source={source} style={computedStyle.image} />;
    }

    if (typeof children === 'string') {
      return <RNText style={computedStyle.text}>{children}</RNText>;
    }

    return children;
  };

  return (
    <RNView style={computedStyle.container} {...rest}>
      {renderContent()}
    </RNView>
  );
};

Avatar.defaultProps = {
  bg: 'transparent',
  size: 48,
  color: 'gray500',
  rounded: 'circle',
  fontSize: 'md',
  shadow: 'none',
  shadowColor: 'gray500',
  position: 'relative',
  zIndex: 1,
};

Avatar.Group = AvatarGroup;

export { Avatar };
