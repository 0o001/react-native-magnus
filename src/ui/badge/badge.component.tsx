import * as React from 'react';
import { useContext } from 'react';
import { View as RNView, Text as RNText } from 'react-native';

import { BadgeProps } from './badge.type';
import { getStyle } from './badge.style';
import { ThemeContext } from '../../theme';
import { Div } from '../div/div.component';

const Badge: React.FunctionComponent<BadgeProps> = (props: BadgeProps) => {
  const {
    h,
    w,
    m,
    mt,
    mr,
    bg,
    mb,
    ml,
    ms,
    p,
    pr,
    pt,
    pb,
    pl,
    minH,
    minW,
    style,
    shadow,
    rounded,
    position,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    alignItems,
    fontSize,
    justifyContent,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderEndWidth,
    shadowColor,
    children,
    right,
    top,
    left,
    bottom,
    color,
    count,
    opacity,
    ...rest
  } = props;
  const { theme } = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  return (
    <Div
      h={[20, 100]}
      w={[20, 100]}
      bg={['red500', 'yellow500']}
      mt={[0, 'md']}
    />
  );

  return (
    <RNView style={computedStyle.container}>
      <RNView style={{ alignSelf: 'flex-start' }}>
        {typeof children !== 'string' && children}
        <RNView style={computedStyle.div} {...rest}>
          {typeof children === 'string' && (
            <RNText style={computedStyle.text}>{children}</RNText>
          )}
        </RNView>
      </RNView>
    </RNView>
  );
};

Badge.defaultProps = {
  bg: 'green700',
  rounded: 'circle',
  shadow: 0,
  color: 'white',
  top: 0,
  right: 0,
  shadowColor: 'gray900',
  position: 'relative',
  fontSize: 'sm',
};

export { Badge };
