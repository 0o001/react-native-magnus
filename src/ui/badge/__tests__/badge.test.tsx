import React from "react";
import renderer from "react-test-renderer";

import { Badge } from "../badge.component";
import { Avatar } from "../../avatar/avatar.component";
import { ThemeProvider } from "../../../theme";

describe("Badge Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Badge>
          <Avatar>1</Avatar>
        </Badge>
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();

    const typeInstance = instance.root.findByType(Avatar);
    expect(typeInstance.props.children).toBe("1");
  });
});
