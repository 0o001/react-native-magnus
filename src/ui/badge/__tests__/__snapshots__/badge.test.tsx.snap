// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component should render without issues 1`] = `
<View
  style={
    Object {
      "minHeight": 10,
      "minWidth": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderRadius": 1000000,
        "height": 48,
        "justifyContent": "center",
        "position": "relative",
        "width": 48,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 15,
          "textAlign": "center",
          "textAlignVertical": "center",
        }
      }
    >
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#3e9a19",
        "borderRadius": 1000000,
        "justifyContent": "center",
        "minHeight": 10,
        "minWidth": 10,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 9999,
      }
    }
  />
</View>
`;
